libs/isograph-react/src/core/optimisticProxy.ts

```diff
@@ -0,0 +1,415 @@
+import {
+  callSubscriptions,
+  insertEmptySetIfMissing,
+  type EncounteredIds,
+} from './cache';
+import type {
+  BaseStoreLayerData,
+  IsographEnvironment,
+  StoreLayerData,
+  StoreLink,
+  StoreRecord,
+} from './IsographEnvironment';
+import { logMessage } from './logging';
+
+export function getOrInsertRecord(
+  dataLayer: StoreLayerData,
+  link: StoreLink,
+): StoreRecord {
+  const recordsById = (dataLayer[link.__typename] ??= {});
+  return (recordsById[link.__link] ??= {});
+}
+
+export function readOptimisticRecord(
+  environment: IsographEnvironment,
+  link: StoreLink,
+): StoreRecord {
+  return new Proxy<StoreRecord>(
+    {},
+    {
+      get(_, p) {
+        let node: StoreLayer | null = environment.store;
+
+        while (node !== null) {
+          const storeRecord = node.data[link.__typename]?.[link.__link];
+          if (storeRecord === null) {
+            return null;
+          }
+          if (storeRecord !== undefined) {
+            const value = Reflect.get(storeRecord, p, storeRecord);
+            if (value !== undefined) {
+              return value;
+            }
+          }
+          node = node.parentStoreLayer;
+        }
+      },
+      has(_, p) {
+        let node: StoreLayer | null = environment.store;
+
+        while (node !== null) {
+          const storeRecord = node.data[link.__typename]?.[link.__link];
+          if (storeRecord != undefined) {
```

Comment:
`!= undefined` should be linted against, tbh, `!=` should only (I think) be used with `null`